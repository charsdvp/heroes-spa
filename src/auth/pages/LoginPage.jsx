import { useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import { AuthContext } from '../context/AuthContext'

export const LoginPage = () => {
  // traemos nuestro contexto
  const { login } = useContext(AuthContext)
  const navigate = useNavigate()

  const onLogin = () => {
    // llamamos nuestra funcion y recibe un nombre -> name
    login('Carlos Sanchez')
    // nos trasladamos a la pagina principal
    navigate('/', {
      replace: true
    })
  }
  return (
    <div className='container mt-5'>
      <h1>Login</h1>
      <button
      className='btn btn-primary'
      onClick={onLogin}>Login</button>
    </div>
  )
}
