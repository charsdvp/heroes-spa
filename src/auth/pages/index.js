// archivo barril
export * from './LoginPage'
