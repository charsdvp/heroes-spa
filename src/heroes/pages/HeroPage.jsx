import { useParams, Navigate } from 'react-router-dom'
import { getHeroById } from '../helpers'
export const HeroPage = () => {
  // useParams es un React Router Hook que le permite acceder a parámetros dinámicos en la URL. useParams devuelve un objeto de pares clave
  const { id } = useParams()
  const hero = getHeroById(id)
  // console.log(hero)
  // redireccionamos a otra pagina si no tenemos un heroe
  if (!hero) {
    return <Navigate to='/'/>
  }

  return (
    <h1>HeroPage</h1>
  )
}
