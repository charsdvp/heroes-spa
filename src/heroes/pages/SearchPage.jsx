import { useLocation, useNavigate } from 'react-router-dom'
import { useForm } from '../hooks/useForm'
import queryString from 'query-string'
import { getHeroesByName } from '../helpers'
import { HeroCard } from '../components/HeroCard'

export const SearchPage = () => {
  const navigate = useNavigate()
  // para manejar los query params
  const location = useLocation() // nos dara nuestra localizacion y nuestros params
  // procesamos nuestros querys
  const { q = '' } = queryString.parse(location.search)
  // obtenemos los heroes filtrados por el query - name
  const heroes = getHeroesByName(q)

  const showSearch = q.length === 0
  const showError = !showSearch && heroes.length === 0

  // reutilizamos nuestro hook useform para controlar el form
  const { searchtext, onInputChange } = useForm({
    searchtext: q
  })
  // cuando enviamos nuestros valores del form
  const onSearchSubmit = (event) => {
    event.preventDefault()
    // limpiamos el texto
    // if (searchtext.trim().length <= 1) return
    // navegamos hacia -> searchtext
    navigate(`?q=${searchtext.toLowerCase().trim()}`)
  }
  return (
    <>
      <h1>SearchPage</h1>
      <div className="row">
        <div className='col-5'>
          <h4>Searching...</h4>
          <form onSubmit={ onSearchSubmit }>
            <input type='text'
              placeholder='Search a hero'
              className='form-control'
              name='searchtext'
              autoComplete='off'
              value={searchtext}
              onChange={onInputChange} />
            <button className='btn btn-outline-primary mt-1'>Search</button>
          </form>
        </div>
        <div className='col-7'>
          <h4>Results</h4>
          <div className='alert alert-primary' style={{ display: showSearch ? '' : 'none' }}>
            Search a hero
          </div>
          <div aria-label='alert-danger' className='alert alert-danger' style={{ display: showError ? '' : 'none' }}>
            No hero with <b>{q}</b>
          </div>
          {
            heroes.map(hero => (
              <HeroCard key={hero.id} {...hero}/>
            ))
          }
        </div>
      </div>
    </>
  )
}
