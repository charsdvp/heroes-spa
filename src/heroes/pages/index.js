// archivo de barril
export * from './DcPage'
export * from './MarvelPage'
