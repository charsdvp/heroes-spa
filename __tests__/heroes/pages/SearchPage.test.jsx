import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { SearchPage } from '../../../src/heroes/pages/SearchPage'

const mockedUseNavigate = jest.fn()
// hacemos un mock de toda nuestra libreria con ...jest y solo modificamos la propiedad que necesitamos
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate
}))

describe('Pruebas en <SearchPage>', () => {

  beforeEach(() => jest.clearAllMocks())

  test('Debe de mostrarse correctamente con valores por defecto', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )
    // screen.debug()
    // tomamos un snapshot
    expect( container ).toMatchSnapshot()
  })
  test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )
    // nuestro input debe de tener los mismo que en nuestro query
    const input = screen.getByRole('textbox')
    expect(input.value).toBe('batman')
    // se debe de renderizar una imagen que contenga el texto batman
    const img = screen.getByRole('img')
    expect(img.src).toContain('dc-batman.jpg')
    // no se deberia de mostrar nuestro alertDanger
    const alertDanger = screen.getByLabelText('alert-danger')
    expect(alertDanger.style.display).toBe('none')
  })
  test('Debe de mostrar un error si no se encuentra el hero',() => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchPage />
      </MemoryRouter>
    )
    // se deberia de mostrar nuestro alertDanger
    const alertDanger = screen.getByLabelText('alert-danger')
    expect(alertDanger.style.display).toBe('')
  })
  test('Debe de llamar el navigate a la pantalla nueva', () =>{
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>
    )
    // el valor que se enviara es superman
    const input = screen.getByRole('textbox')
    fireEvent.change(input, {target: {name: 'searchtext', value: 'superman'}})
    // disparamos el evento del form con el valor de superman
    const form = screen.getByRole('form')
    fireEvent.submit(form)

    expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman')
  })
})