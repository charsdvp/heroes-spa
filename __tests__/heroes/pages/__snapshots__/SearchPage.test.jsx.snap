// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching...
      </h4>
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchtext"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <div
        class="alert alert-primary"
      >
        Search a hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
